tests/posix/common/src/clock.c

```diff
@@ -70,8 +71,31 @@ void test_posix_realtime(void)
 	ret = clock_gettime(CLOCK_REALTIME, &rts);
 	zassert_equal(ret, 0, "Fail to get realtime clock");
 
-	zassert_equal(rts.tv_sec, mts.tv_sec, "Seconds not equal");
-	zassert_equal(rts.tv_nsec, mts.tv_nsec, "Nanoseconds not equal");
+	/* Between both clock_gettime(), time may have been increased by 1ms
+	 * (precision of k_uptime_get()). This could imply a rollover of nsec
+	 * counter and increment number of seconds.
+	 */
+	if ((rts.tv_nsec != mts.tv_nsec) || (rts.tv_sec != mts.tv_sec)) {
+		/* If nsec or sec are not equal, then rts should be equal to
+		 * mts + 1ms
```

Comment:
That comment doesn't sound like correct logic to me.  Nothing in either Zephyr or POSIX guarantees that two consecutive times differ by exactly 1ms.

It also doesn't seem like that's the logic implemented below, which is looking for a delta of exactly one **nanosecond**!  And I can 100% guarantee that Zephyr won't produce that -- it's fully five orders of magnitude more precise than the highest precision tick rate on any existing system.

Reply:
This test is verifying the behavior of the Zephyr POSIX implementation of `clock_gettime()`, which returns the value of `k_uptime_get()` for monotonic and `k_uptime_get()` plus a base offset for "realtime".  So unless/until that changes it will always increment in multiples of 1 ms.

None of which is against the point you or @pfalcon raise about this test's inherent weaknesses, just that it is meeting the minimum expectations for what it's intended to verify.

Reply:
> So unless/until that changes it will always increment in multiples of 1 ms.

The point, that it may, and **does**, increment in **multiples** of 1ms. Many people (re)discover again and again that e.g. on QEMU, time granularity is 10ms (see e.g. cb2bfcb9c0e73e271f45a3d6ac8cc9718810e3c9 and gazillion of other commits (I looked for mine and saw 3-4 from other people on the matter)). Thus, both saying and actually testing that difference can be 1ms is of little help.

Reply:
Here's for example how I do it: https://github.com/zephyrproject-rtos/zephyr/blob/21358baa72cea9c23be57444eb91444774842f97/tests/net/socket/select/src/main.c#L26 (again, defining a symbolic fuzz factor, and set it conservatively ("crudest" known (at least, to me) time granularity is 10ms)).

Reply:
As the test describes in the comment it also uses `k_usleep(1)` to align to the start of a tick, so that the two times will be equal even on QEMU within the limits of the test's goals and context.  In fact, there's only one platform where this patch is asserted to be relevant, and I can't replicate the failure on that platform.

Really what this test is trying to do is to confirm, through a secondary effect, that the offset of `CLOCK_REALTIME` from `CLOCK_MONOTONIC` is zero.  We could just remove the test instead, as even a 1 ms slop as added here invalidates the purpose of the test.

In any case, I confirm that this is a working solution to #25453 if in fact that issue is reproducible.  The logic does not look for a delta of 1 ns, but I agree it's too complex and the name `mts_tvnsec_plus1` is misleading.  So I've retracted my approval, though I'd still approve the alternative implementation I suggested.

Reply:
Sure, I reproduced the issue systematicvally (on a certain commit). And yes my solution solves the issue.
But I understand it is a bit complex, and I will give a try with your suggested implementation.

Reply:
> As the test describes in the comment it also uses k_usleep(1) to align to the start of a tick,

That k_usleep() and comment was added by me 10 days ago on @andyross suggestion, on seeing a failing case very similar to #25453 (that case was also on real hw, not qemu: https://github.com/zephyrproject-rtos/zephyr/issues/25017).

The more strange that this "alignment to a start of a ticket" didn't work and 10 days later, a similar issue pops up on another hw.


I don't comment on other points, because they're similar to what I already said in my review comment: https://github.com/zephyrproject-rtos/zephyr/pull/25454#pullrequestreview-415200109

The only new thought I can bring, is that if there's a desire to keep applying small-scale (1-2 lines, not counting comments) workarounds, then one can wrap 2 clock_gettime() calls in irq_lock(). We still will need to remove that comparison eventually (e.g. when CLOCK_REALTIME will be actually RTC-backed). I didn't remove it 10 days ago only because I consider that we lack a lot of test coverage, so removing tests added by somebody else isn't a good way to encourage people to write more. (Otherwise, we can "resolve" an issue with any test by just removing it.)

Reply:
> I understand it is a bit complex,

It's not complex, it's confusing. It's unclear why workarounds applied so far don't work consistently across hw, and we need to apply more (and then it would make sense to apply such which wouldn't break again in a week).